### PR TITLE
msg/async/dpdk: rebase to spdk/dpdk

### PR DESCRIPTION
**Pull request by liu-chunmei**
this patch make msg/async/dpdk can be built based on spdk/dpdk library since we remove src/dpdk folder.(need build and install dpdk library first).  The ceph cluster will startup, but still need do further functional debug.

need disable some test code since we set -DWITH_TESTS=OFF. 
Signed-off-by: chunmei chunmei.liu@intel.com

### CODE REVIEW

File: cmake/modules/Finddpdk.cmake [hunk 1]
```diff
@@ -13,6 +13,8 @@ find_library(DPDK_rte_kvargs_LIBRARY rte_kvargs)
 find_library(DPDK_rte_mbuf_LIBRARY rte_mbuf)
 find_library(DPDK_rte_ethdev_LIBRARY rte_ethdev)
 find_library(DPDK_rte_mempool_LIBRARY rte_mempool)
+find_library(DPDK_rte_mempool_ring_LIBRARY rte_mempool_ring)
```
Comment (liu-chunmei) [inline cmake/modules/Finddpdk.cmake]: spdk/dpdk separate rte_mempool library to three individual library rte_mempool, rte_mempool_ring, rte_mempool_stack, so need include all of them.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -863,6 +863,9 @@ add_executable(ceph-osd ${ceph_osd_srcs})
 add_dependencies(ceph-osd erasure_code_plugins)
 target_link_libraries(ceph-osd osd os global-static common
   ${BLKID_LIBRARIES} ${RDMA_LIBRARIES})
```
Comment (liu-chunmei) [inline src/CMakeLists.txt]: here, if just link to common (common link to ceph-common-deps), can't link all DPDK_LIBRARIES (don't know why), but do  target link here again it works.

File: src/msg/async/dpdk/DPDK.cc [hunk 3]
```diff
@@ -532,7 +532,7 @@ bool DPDKQueuePair::init_rx_mbuf_pool()
     for (int i = 0; i < mbufs_per_queue_rx; i++) {
       rte_mbuf* m = rte_pktmbuf_alloc(_pktmbuf_pool_rx);
       assert(m);
```
Comment (liu-chunmei) [inline src/msg/async/dpdk/DPDK.cc]: spdk/dpdk changed the refcnt setting check, so need modify it according dpdk library.

File: src/msg/async/dpdk/DPDK.cc [hunk 4]
```diff
@@ -562,7 +562,6 @@ bool DPDKQueuePair::init_rx_mbuf_pool()
     }
   }
```
Comment (liu-chunmei) [inline src/msg/async/dpdk/DPDK.cc]: rte_mempool_count and rte_mempool_free_count are not defined in spdk/dpdk library, just remove it.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 5]
```diff
@@ -52,7 +52,7 @@ namespace dpdk {
     }
 
     bool done = false;
```
Comment (liu-chunmei) [inline src/msg/async/dpdk/dpdk_rte.cc]: this is last mistake,  my debug code is not clean completely.

**Comment by yuyuyu101** [comment]
plz add signed-off-by

**Comment by tchaikov** [comment]
@liu-chunmei i incorporated your change into #18944. at least, it builds fine.

**Comment by liu-chunmei** [comment]
@tchaikov It seems OK, but I think need test, at least to start ceph with DPDK ON. build fine not means all libraries are included (rte_mempool_* libraries and net driver libraries no symbol link relationship with ceph-osd but needed by ceph-osd, that libraries will register something used by ceph-osd, if not register no symbol error but function will have error). I found it this time if I don't relink DPDK_LIBRARY to ceph-osd, some library is not included in ceph-osd binary (at least in my environment). So with my change, ceph can start up but when do fio test still has some assert happen, I am debugging on it now. To make sure your change works, you can start ceph cluster with DPDK ON, then make sure all libraries needed are included in ceph-osd.

**Comment by tchaikov** [comment]
thanks @liu-chunmei ! it has been tested at http://pulpito.ceph.com/kchai-2017-11-17_09:57:45-rados-wip-kefu-testing-2017-11-17-1613-distro-basic-smithi/ . the failure is due to #18974, i guess.

**Comment by tchaikov** [comment]
since #18944 has been merged, closing this one.

**Pull request closed without merging**